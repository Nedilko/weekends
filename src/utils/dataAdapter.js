import { getDefaultSettings } from '@utils/settings'
import {
  writeToLocalstorage,
  readFromLocalstorage,
} from '@utils/localstorageAdapter'

const loadSettings = () => {
  const localStorData = readFromLocalstorage('settings')
  if (!localStorData) {
    writeSettings(getDefaultSettings())
    return getDefaultSettings()
  }
  return JSON.parse(atob(localStorData))
}

const writeSettings = (settings) => {
  writeToLocalstorage('settings', btoa(JSON.stringify(settings)))
}

export { loadSettings, writeSettings }
