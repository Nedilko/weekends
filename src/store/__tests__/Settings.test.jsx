import { render, screen, userEvent } from '@utils/test-utils'
import SettingsContext, { SettingsContextProvider } from '@store/Settings'
import { useContext } from 'react'
import { loadSettings, writeSettings } from '@utils/dataAdapter'

jest.mock('@utils/dataAdapter')

describe('Settings context provider', () => {
  function WrappedComponent() {
    const { data, handleApply } = useContext(SettingsContext)
    return (
      <>
        <div>{JSON.stringify(data, null, '\t')}</div>
        <button
          onClick={() =>
            handleApply({
              hour: 19,
              greetingsText: 'Hello world',
              theme: 'dark',
              useSystemTheme: true,
              isFirstLoad: false,
            })
          }
        >
          Apply
        </button>
      </>
    )
  }

  beforeEach(() => {
    loadSettings.mockImplementation(() => ({
      day: 6,
      hour: 18,
      greetingsText: 'Hello',
      theme: 'light',
      useSystemTheme: false,
      isFirstLoad: true,
    }))
    writeSettings.mockImplementation(() => {})
  })

  afterEach(() => {
    loadSettings.mockClear()
    writeSettings.mockClear()
  })

  it('should load settings', () => {
    render(
      <SettingsContextProvider>
        <WrappedComponent />
      </SettingsContextProvider>
    )
    expect(loadSettings).toHaveBeenCalledTimes(1)
    expect(writeSettings).not.toHaveBeenCalled()
  })

  it('should write settings', async () => {
    const user = userEvent.setup()
    render(
      <SettingsContextProvider>
        <WrappedComponent />
      </SettingsContextProvider>
    )
    await user.click(screen.getByRole('button', { name: /apply/i }))
    expect(loadSettings).toHaveBeenCalledTimes(2)
    expect(writeSettings).toHaveBeenCalledWith({
      day: 6,
      hour: 19,
      greetingsText: 'Hello world',
      theme: 'light',
      useSystemTheme: true,
      isFirstLoad: false,
    })
  })
})
