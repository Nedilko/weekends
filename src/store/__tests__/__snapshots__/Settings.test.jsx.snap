// Vitest Snapshot v1

exports[`Settings context provider > should render all data 1`] = `
<div>
  <div>
    {
	"day": 6,
	"hour": 18,
	"greetingsText": "Hello",
	"theme": "light",
	"useSystemTheme": false,
	"isFirstLoad": true
}
  </div>
  <button>
    Apply
  </button>
</div>
`;

exports[`Settings provider consistance data > should render all data 1`] = `
<div>
  <div>
    {
	"day": 6,
	"hour": 18,
	"greetingsText": "Hello",
	"theme": "light",
	"useSystemTheme": false,
	"isFirstLoad": true
}
  </div>
  <button>
    Apply
  </button>
</div>
`;
